### PR TITLE
feat(Session Replay): Add breadcrumbs to session replay

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Converting session breadcrumbs to replay breadcrumbs

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:76]: timestamp and endTimestamp are the same because the network breadcrumb dont have the start. An improvement for network tracker is coming in the next PR

File: Sources/Sentry/SentryScope.m
Comment (brustolin) [inline Sources/Sentry/SentryScope.m:36]: Moved to SentryScope+Private

File: Sources/Sentry/SentrySerialization.m
Comment (brustolin) [inline Sources/Sentry/SentrySerialization.m:331]:
```suggestion
```

File: Sources/Swift/Extensions/StringExtensions.swift
Comment (philipphofmann) [inline Sources/Swift/Extensions/StringExtensions.swift:49]: `m`: I'm not sure if this is a problem, but if you perform this method on `hello_worldHello`, the outcome is `helloWorldhello`, and I think it should be `helloWorldHello` because we only use uppercase if the previous char was _. This logic doesn't keep existing uppercase chars.

File: Tests/SentryTests/Integrations/SessionReplay/SentryBreadcrumbReplayConverterTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/SessionReplay/SentryBreadcrumbReplayConverterTests.swift:96]: `m`: Also asserting the `startTimesamp` and `endTimestamp` properties of the payload improves this test case, as no other test validates these properties.

File: Tests/SentryTests/Integrations/SessionReplay/SentryReplayRecordingTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/SessionReplay/SentryReplayRecordingTests.swift:52]: `l`: Please ditch Nimble

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:17]: Ah, so great. We finally can use `compactMap` instead of iterating over the list and throwing out the nil values in ObjC. 💯

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:30]: `l`: Does it make sense to define reusable constantans for things like `device.connectivity`, `BATTERY_STATE_CHANGE`, so we ensure both this class and `SentrySystemEventBreadcrumbs` will use the same keywords?

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:37]: `l`: Please comment on why we set this to `nil`. I don't get it.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:44]: `m`: Why do we only keep the level for this crumb here? Shouldn't we also keep the level for all others?

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:34]: `m`: Why do we filter the data here? If we add extra info in the `SentrySystemEventBreadcrumbs`, that would also be useful for SR, the data would be dropped here.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:31]: `m`: Why do we drop crumbs for which the networkType is empty?

File: Tests/SentryTests/Integrations/SessionReplay/SentryBreadcrumbReplayConverterTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/SessionReplay/SentryBreadcrumbReplayConverterTests.swift:32]: `l`: Isn't that a bit cleaner? The same applies to all other event["type"] assertions.
```suggestion
        XCTAssertEqual(event?["type"] as? Int, SentryRRWebEventType.custom.rawValue)
```

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:49]: `l`: This check is redundant. We already check for that in `replayBreadcrumb`. We could pass `timestamp` from `replayBreadcrumb` as a nonnil argument to here to ditch it.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:54]: `m`: Do we really need to check if the position is `portrait` or `landscape`? Can't we just send what the SDK already created?

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:76]: `m`: I'm a bit confused why we convert a breadcrumb to a span event. What made you choose this way?

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:5]: `h`: This class heavily relies on the implementation detail of classes such as `SentryBreadcrumbTracker`, `SentrySystemEventBreadcrumbs`, etc. I'm worried if we change something in our breadcrumbs implementation that we will break the logic here. Do you have suggestions to mitigate this risk?

**Review by philipphofmann — COMMENTED** [review]
I found a couple of issues.

File: Sources/Sentry/SentrySessionReplay.m
Comment (brustolin) [reply Sources/Sentry/SentrySessionReplay.m:302]: Configure scope is thread safe

File: Sources/Swift/Extensions/StringExtensions.swift
Comment (brustolin) [reply Sources/Swift/Extensions/StringExtensions.swift:49]: I dont have a preference. I just followed android implementation,
but I believe the premises is that snake case, [by convention](https://en.wikipedia.org/wiki/Snake_case), dont have upper cases in the middle of the word.

> Snake case (stylized as snake_case) is the naming convention in which each space is replaced with an underscore (_) character, and words are written in lowercase


@romtsn Thoughts?

File: Tests/SentryTests/Integrations/SessionReplay/SentryBreadcrumbReplayConverterTests.swift
Comment (brustolin) [reply Tests/SentryTests/Integrations/SessionReplay/SentryBreadcrumbReplayConverterTests.swift:32]: Its clear, but we can change `SentryRRWebEventType.custom` by mistake and the test will succeed where the server side is expecting 5.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:76]: Because Network information for session replay has a different struct.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:5]: That's a good point, that I dont have an answer for other than we should be aware of it when review PRs.

Session replay is expecting a specific data struct different from the one used by our breadcrumbs, that's why we need the converter, the best approach would be strongly typed breadcrumbs, which we dont have.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:76]: Btw, this is a RRWebSpan, no relation with trace span

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:37]: Removing information that will not be used.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:34]: The extra data is not displayed in the UI, session replay only need data that they have a use for.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:31]: Session replay UI has a specific implementation, any information that is not known for session replay will not be displayed.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:44]: Custom breadcrumbs will be displayed in a different tab in the UI, we send all the information of the breadcrumb in this cases.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:49]: This is more of an unwrapping of timestamp than check if it is nil.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:20]: Thanks for converting this to Swift ❤️

File: Tests/SentryTests/Extensions/StringExtensionsTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Extensions/StringExtensionsTests.swift:5]: 💯 Thanks

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:41]: I'm sorry I haven't been clear enough. I meant, please this comment to the code.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:48]: Please add this info as a comment in the code, as it's not obvious and subsequent readers of the code could ask the same question.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:38]: Again, please add this info as a comment to the code, as it's not obvious.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:35]: Again, please add this info as a comment to the code, as it's not obvious.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift]: Fine by me.

**Review by philipphofmann — COMMENTED** [review]
Almost, LGTM.

**Comment by brustolin** [comment]
> I'm sorry I haven't been clear enough. I meant, please this comment to the code.

Oh no, I did understand, I just do see it. We check some information, if it dont match what we want we ditch it, pretty simple stuff.

But I will add a general comment in the convert function to explain this.

File: Sources/Swift/Extensions/StringExtensions.swift
Comment (romtsn) [reply Sources/Swift/Extensions/StringExtensions.swift:49]: hm, I'm not sure the android implementation is the same, we only uppercase the first character after the underscore, but keep the case of the other characters. So we never lowercase anything

Comment (brustolin) [reply Sources/Swift/Extensions/StringExtensions.swift:49]: Ow, so I got it wrong. Ok, I can change it.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (romtsn) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:5]: Probably nitpicking, but: should we name it `SentryReplayBreadcrumbConverter` rather? Because now it's sounds like we're converting breadcrumb replays and not replay breadcrumbs 😅

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (romtsn) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:16]: also nit: I'd expect the main method of the `XXXConverter` class to be just `convert` 🙂

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (philipphofmann) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:25]: Thanks 👍

**Review by philipphofmann — APPROVED** [review]
Thanks for adding tests to ensure changes in the SDK crumbs don't break SR crumbs. Maybe it would make sense to move all the SR crum tests from the SDK crumb test classes to the SR crumb converter tests class.

**Comment by brustolin** [comment]
> Maybe it would make sense to move all the SR crum tests from the SDK crumb test classes to the SR crumb converter tests class

Some of those tests have a context already in place in their respective test classes. Its easier to leave it as it is.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (romtsn) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:81]: > $0.key == "response_body_size" 

`h`: I think this condition will never be met, because `supportedNetworkData` does not contain `response_body_size`/`request_body_size`. You should probably add these keys to the list

Comment (romtsn) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:81]: and I think in general this logic is slightly off - basically the keys should be `data['requestBodySize']` or `data['responseBodySize']` - I hope it makes more sense this way

Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:81]: But If I got it right from java, you're using `bodySize`
 ```Java
 val breadcrumbData = mutableMapOf<String, Any?>()
            for ((key, value) in breadcrumb.data) {
                if (key in supportedNetworkData) {
                    breadcrumbData[
                        key
                            .replace("content_length", "body_size")
                            .substringAfter(".")
                            .snakeToCamelCase()
                    ] = value
                }
            }
```
Not sure If I got your comment.

Comment (romtsn) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:81]: yeah, I guess I wasn't clear enough - so you have this line:
```
guard supportedNetworkData.contains($0.key) else { return }
```
but because `supportedNetworkData` doesn't contain `response_body_size`/`request_body_size` in it, this will always early-return, so you will never get to the next line for those keys. So you have to update `supportedNetworkData` with those keys.

to the second point - yes we use `bodySize` but here we actually replace a `content_length` substring with `body_size`, so `response_content_length` transforms to `response_body_size` and then we camelCase it to `responseBodySize`. I hope it makes it more clear

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (vaind) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:29]: For the RN SDK (and Flutter later), there should be a way to extend or override this so that it can implement mappings for its own breadcrumbs. AFAIU there's currently no way to instantiate `SentrySessionReplay` from a hybrid SDK so it can't provide its own implementation. Or has anything changed in that regard lately? It came up in the discussion in [this issue](https://github.com/getsentry/sentry-cocoa/issues/2616) but I haven't noticed any updates.

Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:29]: Nothing new yet. Still need to find a proper way to make SentrySessionReplay Hybrid public

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (romtsn) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:65]: alright, I guess we have similar problems for Fragments, but I'll leave it up to you to decide. We do the filtering here: https://github.com/getsentry/sentry-java/pull/3406/files#diff-2f6abcfaf76542e5e58a9fbd4458b23ca658763d9bc0f989abfdeeee5c5898ddR239

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (romtsn) [inline Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:89]:
```suggestion
    private func getLevel(breadcrumb: Breadcrumb) -> SentryLevel {
```
swiftlint please :P

**Review by romtsn — APPROVED** [review]
LGTM!

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:89]: We do have one in place with git hooks. Not sure what is going on here.

File: Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift
Comment (vaind) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:29]: Alright, with breadcrumbs this starts to be a blocker for react-native too so when do you think you could have a look into that?

Comment (brustolin) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift:29]: I guess @bruno-garcia can help us with the priorities. 
Right now Im working with touch tracker. 
I can work on exposing SessionReplay for hybrids after it, or I can give it a pause right now.

Comment (bruno-garcia) [reply Sources/Swift/Integrations/SessionReplay/SentryBreadcrumbReplayConverter.swift]: I'd say it's OK to push multi touch in order to unblock a react native breadcrumbs. We need to get that one out

**Pull request merged**
